### PR TITLE
added embedded field handling

### PR DESCRIPTION
**Pull request by jeffDuce**

### CODE REVIEW

File: xop.go [hunk 1]
```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
...
+	tag = parts[len(parts) - 1]
+
+	// return the XMLName from the front of the remaining tag
+	return strings.Split(tag, ",")[0]
+}
+
+// getExplicitXMLName gets the xml name which is explicitly set in the xml tag on the XMLName field
+func getExplicitXMLName(val reflect.Value) string {
+	// only a value of type reflect.Struct can have an XMLName field
+	if val.Type().Kind() == reflect.Struct {
```
Comment (rmrobinson-textnow) [inline xop.go]: I personally prefer the early return approach, i.e. use:
```
if val.Type().Kind() != reflect.Struct {
  return ""
}
... other code ...
``` 
as it keeps the level of indentation down as you get into the body of the function.

File: xop.go [hunk 2]
```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
...
-		rXMLTypeField := rXMLName.Type().Field(i)
-		rXMLValueField := rXMLName.Field(i)
+	// the value has been unwrapped
+	return val
+}
 
-		if rXMLTypeField.Name == "Local" && rXMLValueField.String() != path[1] {
-			return reflect.Value{}, fmt.Errorf("invalid XML object path, expected %s, got %s", rXMLValueField.String(), path[1])
-		}
+func getNameFromTag(tag string) string {
```
Comment (rmrobinson-textnow) [inline xop.go:194]: It may be useful to add a comment here with an example field you're trying to handle.

File: xop.go [hunk 3]
```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
...
+		// skip omitted fields
+		if tag == "-" {
+			continue
+		}
+
+		// unwrap the value
+		valueField = unwrapValue(valueField)
+
+		// check if the value was unwrapped completely
+		if valueField.Type().Kind() == reflect.Slice || valueField.Type().Kind() == reflect.Slice || valueField.Type().Kind() == reflect.Ptr {
```
Comment (rmrobinson-textnow) [inline xop.go]: I think you're checking whether it's a Slice twice in this conditional? That seems unnecessary?

Comment (rmrobinson-textnow) [reply xop.go]: Did you mean to check for Slice, Array and Ptr?

File: xop.go [hunk 4]
```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
...
+			// if valueField is in path
+			if getNameFromTag(tag) == path[0] {
+				// if valueField is the desired field, return
+				if len(path) == 1 {
+					return valueField, nil
+				}
+
+				// valueField is empty, there is nothing more to search
```
Comment (rmrobinson-textnow) [inline xop.go:118]: I'm not sure I understand this case. Why is valueField empty if len(path) != 1? What if it's >1?

File: xop.go [hunk 5]
```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
...
+
+			// valueField is not in path
+			continue
+		}
+
+		// if the field is an embedded struct, search its fields
+		if typeField.Anonymous {
+			result, err := getFieldFromPath(valueField, path)
+			if err == nil {
+				return result, err
```
Comment (rmrobinson-textnow) [inline xop.go]: returning result, nil here may make it easier to pick out as a success case?

**Review by rmrobinson-textnow — COMMENTED** [review]
This is looking great! A few questions to consider.

File: xop.go (hunk 3 above)
Comment (jeffDuce) [reply xop.go]: yes, I meant to check for Slice, Array and Ptr

File: xop.go (hunk 4 above)
Comment (jeffDuce) [reply xop.go:118]: `unwrapValue` returns an array or slice only if it has no elements, and therefore cannot be unwrapped further. It similarly returns a pointer only if it is a nil pointer. At this point we've unwrapped valueField and found that the result is either an array, a slice or a pointer.

If valueField is the last value in path then it is the field we've been searching for, so we return it. If len(path) > 1, then valueField is not the last value in path, so we return `errFieldNotFound`.

File: xop.go (hunk 2 above)
Comment (jeffDuce) [reply xop.go:194]: I added a comment and a test case

**Review by rmrobinson-textnow — APPROVED** [review]

**Pull request merged**